moshi/src/main/java/com/squareup/moshi/JsonReader.kt

```diff
@@ -0,0 +1,672 @@
+      else -> throw IllegalStateException("Expected a value but was ${peek()} at path $path")
+    }
+  }
+
+  /**
+   * Returns a new `JsonReader` that can read data from this `JsonReader` without
+   * consuming it. The returned reader becomes invalid once this one is next read or closed.
+   *
+   * For example, we can use `peekJson()` to lookahead and read the same data multiple
+   * times.
+   *
+   * ```java
+   * Buffer buffer = new Buffer();
+   * buffer.writeUtf8("[123, 456, 789]")
+   *
+   * JsonReader jsonReader = JsonReader.of(buffer);
+   * jsonReader.beginArray();
+   * jsonReader.nextInt(); // Returns 123, reader contains 456, 789 and ].
+   *
+   * JsonReader peek = reader.peekJson();
+   * peek.nextInt() // Returns 456.
+   * peek.nextInt() // Returns 789.
+   * peek.endArray()
+   *
+   * jsonReader.nextInt() // Returns 456, reader contains 789 and ].
+   * ```
+   */
+  @CheckReturnValue
+  public abstract fun peekJson(): JsonReader
+
+  /**
+   * Returns a [JsonPath](http://goessner.net/articles/JsonPath/) to the current location
+   * in the JSON value.
+   */
+  @get:CheckReturnValue
+  public val path: String
+    get() = getPath(stackSize, scopes, pathNames, pathIndices)
+
+  /** Returns the tag value for the given class key. */
+  @CheckReturnValue
+  public fun <T : Any> tag(clazz: Class<T>): T? {
+    @Suppress("UNCHECKED_CAST")
+    return tags?.let { it[clazz] as T? }
+  }
+
+  /** Assigns the tag value using the given class key and value. */
+  public fun <T : Any> setTag(clazz: Class<T>, value: T) {
+    require(clazz.isAssignableFrom(value.javaClass)) { "Tag value must be of type ${clazz.name}" }
+    val tagsToUse = tags ?: LinkedHashMap<Class<*>, Any>().also { tags = it }
+    tagsToUse[clazz] = value
+  }
+
+  /**
+   * Changes the reader to treat the next name as a string value. This is useful for map adapters so
+   * that arbitrary type adapters can use [nextString] to read a name value.
+   *
+   * In this example, calling this method allows two sequential calls to [nextString]:
+   *
+   * ```java
+   * JsonReader reader = JsonReader.of(new Buffer().writeUtf8("{\"a\":\"b\"}"));
+   * reader.beginObject();
+   * reader.promoteNameToValue();
+   * assertEquals("a", reader.nextString());
+   * assertEquals("b", reader.nextString());
+   * reader.endObject();
+   * ```
+   */
+  @Throws(IOException::class)
+  public abstract fun promoteNameToValue()
+
+  /**
+   * A set of strings to be chosen with [selectName] or [selectString]. This prepares
+   * the encoded values of the strings so they can be read directly from the input source.
+   */
+  public class Options private constructor(
+    internal val strings: Array<out String>,
+    internal val doubleQuoteSuffix: OkioOptions
+  ) {
+    /** Returns a copy of this [Option's][Options] strings. */
+    public fun strings(): List<String> {
+      return buildList(strings.size) {
+        for (string in strings) {
+          add(string)
+        }
+      }
```

Comment:
couldn't you just say `strings.toList()` or ` strings.map { it }`?

Reply:
Collection builders return (truly) immutable instances